File: frontend/src/api/types/schedule.ts
/**
 * Schedule endpoints exactly as documented by the OpenAPI components
 * (backend/app/api/schemas/{schedule,schedule_views}.py): the active plan, versions,
 * generation / approval workflow, Gantt + day views, comparison and continuous replanning.
 */

import type { OrderStatus, ProcessType, ScheduleStatus } from "./enums";
import type { LockResponse, PageResponse } from "./api";
import type { TimeWindow } from "./models";
import type { Bottleneck, ExecutiveKpis, ScheduleEntry, ScheduleMetrics, ScheduleQuality, UnscheduledItem } from "./results";
import type { CapacityReport } from "./analytics";

// --------------------------------------------------------------- versions

/** ScheduleVersionResponse: spec Phase 37 "Schedule v124 — Generated, Algorithm, Configuration, Status". */
export interface ScheduleVersionResponse {
  schedule_version_id: string;
  version_number: number;
  status: ScheduleStatus;
  label: string | null;
  algorithm: string;
  algorithm_version: string;
  profile_id: string;
  profile_version: number;
  config_version: number;
  generated_by: string | null;
  generated_at: string;
  horizon_start: string;
  horizon_end: string;
  run_id: string | null;
  input_snapshot_id: string | null;
  approved_by: string | null;
  approved_at: string | null;
  published_by: string | null;
  published_at: string | null;
  superseded_at: string | null;
  entry_count: number;
  quality_score: number | null;
  quality_summary: string | null;
  metrics: ScheduleMetrics;
  /** 'manual', 'replan:<trigger>' … */
  trigger: string | null;
  previous_version: number | null;
  notes: string | null;
}

export interface AlertCounts {
  total: number;
  by_severity: Record<string, number>;
}

/** VersionAnalyticsResponse: analytics computed together with the version. */
export interface VersionAnalytics {
  as_of: string;
  run_id: string | null;
  kpis: ExecutiveKpis;
  capacity: CapacityReport;
  bottlenecks: Bottleneck[];
  data_quality: Record<string, unknown>;
  alerts: AlertCounts;
  timings: Record<string, number>;
}

export interface WritebackReceipt {
  receipt_id: string;
  mode: string;
  status: string;
  attempted_at: string;
  entries_published: number;
  approved_by: string | null;
  message: string;
  run_id: string | null;
  details: Record<string, unknown>;
}

/** ScheduleVersionDetailResponse: GET /schedule/versions/{v}, POST /schedule/approve|reject. */
export interface ScheduleVersionDetail extends ScheduleVersionResponse {
  quality: ScheduleQuality | null;
  unscheduled: UnscheduledItem[];
  warnings: string[];
  analytics: VersionAnalytics | null;
  writeback_receipt: WritebackReceipt | null;
  details: Record<string, unknown>;
}

export interface ScheduleVersionsQuery {
  page?: number;
  page_size?: number;
  status?: ScheduleStatus;
}

// ------------------------------------------------------------ active plan

/** Which version GET /schedule answers with: latest published, else approved, else draft; none when empty. */
export type PlanStatus = "published" | "approved" | "draft" | "none";

/** SchedulePlanResponse: the active plan with one page of its entries. */
export interface SchedulePlan {
  status: PlanStatus | string;
  version: ScheduleVersionResponse | null;
  entries: PageResponse<ScheduleEntry>;
}

/** Query of GET /schedule and GET /schedule/versions/{v}/entries. */
export interface ScheduleEntriesQuery {
  page?: number;
  /** 1..500 (default 100). */
  page_size?: number;
  /** Repeatable. */
  machine_id?: string | string[];
  start?: string;
  end?: string;
  order_id?: string;
  customer_id?: string;
}

// ------------------------------------------------------------- generation

export interface GenerateRequest {
  /** Stored with the version. */
  note?: string | null;
}

export interface SnapshotInfo {
  snapshot_id: string;
  as_of: string;
  source: string;
  size_bytes: number;
  sha256: string;
  summary: Record<string, number>;
}

/** OptimizationRunResponse: observability record of one engine run. */
export interface OptimizationRun {
  run_id: string;
  kind: string;
  status: string;
  started_at: string;
  finished_at: string | null;
  duration_seconds: number | null;
  orders_considered: number;
  orders_scheduled: number;
  orders_blocked: number;
  objective_score: number | null;
  quality_score: number | null;
  algorithm: string;
  algorithm_version: string;
  profile_id: string | null;
  profile_version: number | null;
  config_version: number | null;
  input_snapshot_id: string | null;
  triggered_by: string | null;
  trigger_reason: string | null;
  error_message: string | null;
  metrics: Record<string, unknown>;
  warnings: string[];
}

/** ScheduleGenerateResponse: POST /schedule/generate (201). */
export interface ScheduleGenerateResponse {
  version: ScheduleVersionDetail;
  run: OptimizationRun;
  snapshot: SnapshotInfo;
  previous_version: number | null;
  priority_results: number;
  entries: number;
  unscheduled: number;
  data_quality_issues: number;
  alerts: number;
  timings: Record<string, number>;
}

// --------------------------------------------------------------- workflow

/** VersionActionRequest: body of POST /schedule/approve, /publish and /reject. */
export interface VersionActionRequest {
  version: number;
  /** Mandatory, audited. */
  reason: string;
}

export interface PublishResponse {
  version: ScheduleVersionDetail;
  receipt: WritebackReceipt;
  superseded: number;
}

// ------------------------------------------------------------- comparison

/** MetricPairResponse: "On-time delivery: 87% → 94%". */
export interface MetricPair {
  label: string;
  before: number | null;
  after: number | null;
  delta: number | null;
}

export type EntryChangeKind = "moved" | "added" | "removed" | "unchanged" | string;

export interface EntryChange {
  operation_id: string;
  order_id: string;
  kind: EntryChangeKind;
  previous_machine_id: string | null;
  proposed_machine_id: string | null;
  previous_start: string | null;
  proposed_start: string | null;
  shift_minutes: number | null;
  reason: string;
}

export interface ScheduleChanges {
  moved_entries: number;
  added_entries: number;
  removed_entries: number;
  unchanged_entries: number;
  changed_orders: string[];
  frozen_violations: number;
  entries: EntryChange[];
}

/** ScheduleComparisonResponse: GET /schedule/compare?a=&b= (spec Phase 36). */
export interface ScheduleComparison {
  a: ScheduleVersionResponse;
  b: ScheduleVersionResponse;
  /** Keyed by metric (on_time_pct, avg_lateness_hours, overall_utilization_pct, total_setup_hours, late_orders, …). */
  metrics: Record<string, MetricPair>;
  quality: MetricPair;
  changes: ScheduleChanges;
  moved_orders: number;
  summary: string;
}

/** ScheduleQualityReportResponse: GET /analytics/schedule-quality. */
export interface ScheduleQualityReport {
  version: ScheduleVersionResponse | null;
  quality: ScheduleQuality | null;
  metrics: ScheduleMetrics | null;
  newest_draft: ScheduleVersionResponse | null;
  /** Active plan vs the newest draft, when both exist. */
  comparison: ScheduleComparison | null;
}

// ---------------------------------------------------------- machine views

/** GanttBlockResponse: one job on the machine board (setup block + run block). */
export interface GanttBlock {
  entry: ScheduleEntry;
  order_id: string;
  customer_id: string | null;
  customer_name: string | null;
  part_id: string | null;
  part_name: string | null;
  order_status: OrderStatus | null;
  setup_start: string;
  start: string;
  end: string;
  locked: boolean;
  late: boolean;
}

/** MachineRowResponse: one machine of the Gantt / day view with its blocks, downtime and locks. */
export interface MachineRow {
  machine_id: string;
  machine_name: string;
  machine_group: string;
  process_type: ProcessType;
  status: string;
  busy_hours: number;
  blocks: GanttBlock[];
  downtime: TimeWindow[];
  locks: LockResponse[];
}

/** GanttResponse: GET /schedule/gantt. */
export interface GanttView {
  version: ScheduleVersionResponse;
  axis_start: string;
  axis_end: string;
  machine_group: string | null;
  process_type: ProcessType | null;
  entries: number;
  rows: MachineRow[];
}

export interface GanttQuery {
  version?: number;
  /** Default: now (or the horizon start). */
  start?: string;
  /** Default: start + 7 days. */
  end?: string;
  machine_group?: string;
  process_type?: ProcessType;
  /** Repeatable. */
  machine_id?: string[];
}

/** DayScheduleResponse: GET /schedule/{date} (plant-local day). */
export interface DaySchedule {
  /** yyyy-MM-dd */
  date: string;
  timezone: string;
  start: string;
  end: string;
  version: ScheduleVersionResponse;
  entries: number;
  machines: MachineRow[];
}

// ------------------------------------------------------------- replanning

export type ReplanTriggerType =
  | "new_order"
  | "order_completed"
  | "machine_down"
  | "machine_up"
  | "material_arrived"
  | "quality_failure"
  | "rework"
  | "production_delay"
  | "customer_priority_change"
  | "manual"
  | "config_change"
  | "scheduled";

export interface ReplanRequest {
  trigger?: ReplanTriggerType;
  reason?: string | null;
}

export interface ReplanEvent {
  type: ReplanTriggerType;
  entity_type: string;
  entity_id: string;
  occurred_at: string;
  message: string;
  order_id: string | null;
  machine_id: string | null;
  details: Record<string, unknown>;
}

export interface ReplanDecision {
  should_replan: boolean;
  reason: string;
  improvement_pct: number;
  changed_entries: number;
  frozen_violations: number;
  requires_approval: boolean;
  triggers: string[];
}

export type ReplanAction = "not_triggered" | "rejected" | "awaiting_approval" | "approved" | "published" | string;

/** ReplanResponse: POST /schedule/replan (spec Phase 11: compare old vs new, approval). */
export interface ReplanOutcome {
  trigger: ReplanTriggerType;
  evaluated_at: string;
  triggered: boolean;
  action: ReplanAction;
  reason: string;
  event_types: string[];
  events: ReplanEvent[];
  decision: ReplanDecision | null;
  active_version: ScheduleVersionResponse | null;
  candidate_version: ScheduleVersionResponse | null;
  comparison: ScheduleComparison | null;
  alert_id: string | null;
}

File: backend/app/api/schemas/schedule.py
"""Schedule schemas: versions, generation, approval workflow, runs and comparison (spec Phases 36, 37)."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.api.schemas.analytics import BottleneckResponse, CapacityResponse, KpisResponse
from app.api.schemas.common import PageResponse, ReasonBody
from app.api.schemas.orders import ScheduleEntryResponse
from app.db.records import OptimizationRunRecord, ScheduleVersionInfo, SnapshotInfo
from app.domain.enums import ScheduleStatus
from app.domain.results import ScheduleMetrics, ScheduleQuality
from app.integration.writeback import WritebackReceipt
from app.services.analytics_service import ScheduleComparison, ScheduleQualityView
from app.services.schedule_service import CurrentPlan, GenerationSummary, PublishOutcome, RunDetails
from app.services.writeback_service import receipt_of

# ------------------------------------------------------------------ versions


class ScheduleMetricsResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    scheduled_orders: int = 0
    unscheduled_orders: int = 0
    scheduled_operations: int = 0
    on_time_orders: int = 0
    late_orders: int = 0
    orders_at_risk: int = 0
    on_time_pct: float = 0.0
    avg_lateness_hours: float = 0.0
    max_lateness_hours: float = 0.0
    total_tardiness_hours: float = 0.0
    total_run_hours: float = 0.0
    total_setup_hours: float = 0.0
    setup_count: int = 0
    makespan_hours: float = 0.0
    overall_utilization_pct: float = 0.0
    machine_utilization_pct: dict[str, float] = {}
    revenue_scheduled: float = 0.0
    revenue_at_risk: float = 0.0
    margin_at_risk: float = 0.0
    wip_orders_avg: float = 0.0

    @classmethod
    def from_domain(cls, m: ScheduleMetrics) -> ScheduleMetricsResponse:
        from app.db.snapshot_codec import to_jsonable

        return cls.model_validate(to_jsonable(m))


class ScheduleQualityResponse(BaseModel):
    """Spec Phase 36: "Schedule Quality: 89/100" with its components."""

    model_config = ConfigDict(extra="ignore")
    score: float
    components: dict[str, float]
    weights: dict[str, float]
    summary: str

    @classmethod
    def from_domain(cls, q: ScheduleQuality) -> ScheduleQualityResponse:
        return cls(score=q.score, components=dict(q.components), weights=dict(q.weights), summary=q.summary)


class UnscheduledItemResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    order_id: str
    operation_id: str | None = None
    reason_code: str
    reason: str
    readiness: str | None = None


class WritebackReceiptResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    receipt_id: str
    mode: str
    status: str
    attempted_at: datetime
    entries_published: int = 0
    approved_by: str | None = None
    message: str = ""
    run_id: str | None = None
    details: dict[str, Any] = {}

    @classmethod
    def from_domain(cls, r: WritebackReceipt) -> WritebackReceiptResponse:
        return cls.model_validate(r.to_dict())


class ScheduleVersionResponse(BaseModel):
    """Spec Phase 37: "Schedule v124 — Generated, Algorithm, Configuration, Status"."""

    model_config = ConfigDict(extra="forbid")
    schedule_version_id: str
    version_number: int
    status: ScheduleStatus
    label: str | None = None
    algorithm: str
    algorithm_version: str
    profile_id: str
    profile_version: int
    config_version: int
    generated_by: str | None = None
    generated_at: datetime
    horizon_start: datetime
    horizon_end: datetime
    run_id: str | None = None
    input_snapshot_id: str | None = None
    approved_by: str | None = None
    approved_at: datetime | None = None
    published_by: str | None = None
    published_at: datetime | None = None
    superseded_at: datetime | None = None
    entry_count: int
    quality_score: float | None = None
    quality_summary: str | None = None
    metrics: ScheduleMetricsResponse
    trigger: str | None = Field(default=None, description="'manual', 'replan:<trigger>' ...")
    previous_version: int | None = None
    notes: str | None = None

    @classmethod
    def from_record(cls, v: ScheduleVersionInfo) -> ScheduleVersionResponse:
        quality = v.quality or {}
        trigger = v.details.get("trigger")
        previous = v.details.get("previous_version")
        return cls(
            schedule_version_id=v.schedule_version_id,
            version_number=v.version_number,
            status=v.status,
            label=v.label,
            algorithm=v.algorithm,
            algorithm_version=v.algorithm_version,
            profile_id=v.profile_id,
            profile_version=v.profile_version,
            config_version=v.config_version,
            generated_by=v.generated_by,
            generated_at=v.generated_at,
            horizon_start=v.horizon_start,
            horizon_end=v.horizon_end,
            run_id=v.run_id,
            input_snapshot_id=v.input_snapshot_id,
            approved_by=v.approved_by,
            approved_at=v.approved_at,
            published_by=v.published_by,
            published_at=v.published_at,
            superseded_at=v.superseded_at,
            entry_count=v.entry_count,
            quality_score=quality.get("score"),
            quality_summary=quality.get("summary"),
            metrics=ScheduleMetricsResponse.model_validate(v.metrics or {}),
            trigger=str(trigger) if trigger is not None else None,
            previous_version=int(previous) if isinstance(previous, int) else None,
            notes=v.notes,
        )


class AlertCountsResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    total: int = 0
    by_severity: dict[str, int] = {}


class VersionAnalyticsResponse(BaseModel):
    """The analytics computed together with the version (stored ``analytics`` JSON)."""

    model_config = ConfigDict(extra="ignore")
    as_of: datetime
    run_id: str | None = None
    kpis: KpisResponse
    capacity: CapacityResponse
    bottlenecks: list[BottleneckResponse]
    data_quality: dict[str, Any] = {}
    alerts: AlertCountsResponse = AlertCountsResponse()
    timings: dict[str, float] = {}


class ScheduleVersionDetailResponse(ScheduleVersionResponse):
    quality: ScheduleQualityResponse | None = None
    unscheduled: list[UnscheduledItemResponse] = []
    warnings: list[str] = []
    analytics: VersionAnalyticsResponse | None = None
    writeback_receipt: WritebackReceiptResponse | None = None
    details: dict[str, Any] = {}

    @classmethod
    def from_record(cls, v: ScheduleVersionInfo) -> ScheduleVersionDetailResponse:
        base = ScheduleVersionResponse.from_record(v).model_dump()
        receipt = receipt_of(v)
        return cls(
            **base,
            quality=ScheduleQualityResponse.model_validate(v.quality) if v.quality else None,
            unscheduled=[UnscheduledItemResponse.model_validate(u) for u in v.unscheduled],
            warnings=list(v.warnings),
            analytics=VersionAnalyticsResponse.model_validate(v.analytics) if v.analytics else None,
            writeback_receipt=WritebackReceiptResponse.model_validate(receipt) if receipt else None,
            details={k: val for k, val in v.details.items() if k != "writeback_receipt"},
        )


class SchedulePlanResponse(BaseModel):
    """``GET /schedule``: the active plan (which status it has) with a page of its entries."""

    model_config = ConfigDict(extra="forbid")
    status: str = Field(description="published / approved / draft / none")
    version: ScheduleVersionResponse | None = None
    entries: PageResponse[ScheduleEntryResponse]

    @classmethod
    def build(cls, plan: CurrentPlan, entries: PageResponse[ScheduleEntryResponse]) -> SchedulePlanResponse:
        return cls(
            status=plan.status,
            version=ScheduleVersionResponse.from_record(plan.version) if plan.version else None,
            entries=entries,
        )


# ---------------------------------------------------------------- generation


class GenerateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    note: str | None = Field(default=None, max_length=2000, description="Stored with the version")


class SnapshotInfoResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    snapshot_id: str
    as_of: datetime
    source: str
    size_bytes: int
    sha256: str
    summary: dict[str, int]

    @classmethod
    def from_record(cls, s: SnapshotInfo) -> SnapshotInfoResponse:
        return cls(
            snapshot_id=s.snapshot_id,
            as_of=s.as_of,
            source=s.source,
            size_bytes=s.size_bytes,
            sha256=s.sha256,
            summary=dict(s.summary),
        )


class OptimizationRunResponse(BaseModel):
    """Observability record of one engine run (spec OBSERVABILITY)."""

    model_config = ConfigDict(extra="forbid")
    run_id: str
    kind: str
    status: str
    started_at: datetime
    finished_at: datetime | None = None
    duration_seconds: float | None = None
    orders_considered: int
    orders_scheduled: int
    orders_blocked: int
    objective_score: float | None = None
    quality_score: float | None = None
    algorithm: str
    algorithm_version: str
    profile_id: str | None = None
    profile_version: int | None = None
    config_version: int | None = None
    input_snapshot_id: str | None = None
    triggered_by: str | None = None
    trigger_reason: str | None = None
    error_message: str | None = None
    metrics: dict[str, Any] = {}
    warnings: list[str] = []

    @classmethod
    def from_record(cls, r: OptimizationRunRecord) -> OptimizationRunResponse:
        return cls(
            run_id=r.run_id,
            kind=r.kind,
            status=r.status,
            started_at=r.started_at,
            finished_at=r.finished_at,
            duration_seconds=r.duration_seconds,
            orders_considered=r.orders_considered,
            orders_scheduled=r.orders_scheduled,
            orders_blocked=r.orders_blocked,
            objective_score=r.objective_score,
            quality_score=r.quality_score,
            algorithm=r.algorithm,
            algorithm_version=r.algorithm_version,
            profile_id=r.profile_id,
            profile_version=r.profile_version,
            config_version=r.config_version,
            input_snapshot_id=r.input_snapshot_id,
            triggered_by=r.triggered_by,
            trigger_reason=r.trigger_reason,
            error_message=r.error_message,
            metrics=dict(r.metrics),
            warnings=list(r.warnings),
        )


class ScheduleGenerateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    version: ScheduleVersionDetailResponse
    run: OptimizationRunResponse
    snapshot: SnapshotInfoResponse
    previous_version: int | None = None
    priority_results: int
    entries: int
    unscheduled: int
    data_quality_issues: int
    alerts: int
    timings: dict[str, float]

    @classmethod
    def from_summary(cls, g: GenerationSummary) -> ScheduleGenerateResponse:
        return cls(
            version=ScheduleVersionDetailResponse.from_record(g.version),
            run=OptimizationRunResponse.from_record(g.run),
            snapshot=SnapshotInfoResponse.from_record(g.snapshot),
            previous_version=g.previous_version,
            priority_results=g.priority_results,
            entries=g.entries,
            unscheduled=g.unscheduled,
            data_quality_issues=g.data_quality_issues,
            alerts=g.alerts,
            timings=dict(g.timings),
        )


class RunDetailsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    run: OptimizationRunResponse
    version: ScheduleVersionResponse | None = None
    snapshot: SnapshotInfoResponse | None = None
    priority_results: int
    data_quality_issues: int

    @classmethod
    def from_domain(cls, d: RunDetails) -> RunDetailsResponse:
        return cls(
            run=OptimizationRunResponse.from_record(d.run),
            version=ScheduleVersionResponse.from_record(d.version) if d.version else None,
            snapshot=SnapshotInfoResponse.from_record(d.snapshot) if d.snapshot else None,
            priority_results=d.priority_results,
            data_quality_issues=d.data_quality_issues,
        )


# ------------------------------------------------------------------ workflow


class VersionActionRequest(ReasonBody):
    version: int = Field(ge=1, description="Schedule version number")


class PublishResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    version: ScheduleVersionDetailResponse
    receipt: WritebackReceiptResponse
    superseded: int

    @classmethod
    def from_domain(cls, p: PublishOutcome) -> PublishResponse:
        return cls(
            version=ScheduleVersionDetailResponse.from_record(p.version),
            receipt=WritebackReceiptResponse.from_domain(p.receipt),
            superseded=p.superseded,
        )


# ---------------------------------------------------------------- comparison


class MetricPairResponse(BaseModel):
    """ "On-time delivery: 87% → 94%"."""

    model_config = ConfigDict(extra="forbid")
    label: str
    before: float | None = None
    after: float | None = None
    delta: float | None = None


class EntryChangeResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    operation_id: str
    order_id: str
    kind: str
    previous_machine_id: str | None = None
    proposed_machine_id: str | None = None
    previous_start: datetime | None = None
    proposed_start: datetime | None = None
    shift_minutes: float | None = None
    reason: str


class ScheduleChangesResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    moved_entries: int
    added_entries: int
    removed_entries: int
    unchanged_entries: int
    changed_orders: list[str]
    frozen_violations: int
    entries: list[EntryChangeResponse] = []


class ScheduleComparisonResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    a: ScheduleVersionResponse
    b: ScheduleVersionResponse
    metrics: dict[str, MetricPairResponse]
    quality: MetricPairResponse
    changes: ScheduleChangesResponse
    moved_orders: int
    summary: str

    @classmethod
    def from_domain(cls, c: ScheduleComparison) -> ScheduleComparisonResponse:
        return cls(
            a=ScheduleVersionResponse.from_record(c.a),
            b=ScheduleVersionResponse.from_record(c.b),
            metrics={k: MetricPairResponse.model_validate(v) for k, v in c.metrics.items()},
            quality=MetricPairResponse.model_validate(c.quality),
            changes=ScheduleChangesResponse.model_validate(c.changes),
            moved_orders=c.moved_orders,
            summary=c.summary,
        )


class ScheduleQualityReportResponse(BaseModel):
    """``GET /analytics/schedule-quality``: active plan quality, compared with the newest draft."""

    model_config = ConfigDict(extra="forbid")
    version: ScheduleVersionResponse | None = None
    quality: ScheduleQualityResponse | None = None
    metrics: ScheduleMetricsResponse | None = None
    newest_draft: ScheduleVersionResponse | None = None
    comparison: ScheduleComparisonResponse | None = None

    @classmethod
    def from_view(cls, v: ScheduleQualityView) -> ScheduleQualityReportResponse:
        return cls(
            version=ScheduleVersionResponse.from_record(v.version) if v.version else None,
            quality=ScheduleQualityResponse.from_domain(v.quality) if v.quality else None,
            metrics=ScheduleMetricsResponse.from_domain(v.metrics) if v.metrics else None,
            newest_draft=ScheduleVersionResponse.from_record(v.draft) if v.draft else None,
            comparison=ScheduleComparisonResponse.from_domain(v.comparison) if v.comparison else None,
        )


__all__ = [
    "AlertCountsResponse",
    "EntryChangeResponse",
    "GenerateRequest",
    "MetricPairResponse",
    "OptimizationRunResponse",
    "PublishResponse",
    "RunDetailsResponse",
    "ScheduleChangesResponse",
    "ScheduleComparisonResponse",
    "ScheduleGenerateResponse",
    "ScheduleMetricsResponse",
    "SchedulePlanResponse",
    "ScheduleQualityReportResponse",
    "ScheduleQualityResponse",
    "ScheduleVersionDetailResponse",
    "ScheduleVersionResponse",
    "SnapshotInfoResponse",
    "UnscheduledItemResponse",
    "VersionActionRequest",
    "VersionAnalyticsResponse",
    "WritebackReceiptResponse",
]
